File: app/scala-ms/financial-accounting-service/src/main/scala/com/example/financial/adapter/repository/AccountRepositoryImpl.scala
package com.example.financial.adapter.repository

import cats.effect.IO
import scalikejdbc.*
import com.example.financial.application.AccountRepository
import com.example.financial.domain.Account
import com.example.common.domain.*
import java.time.LocalDateTime

/**
 * 勘定科目リポジトリの ScalikeJDBC 実装
 */
class AccountRepositoryImpl extends AccountRepository:

  override def findAll(): IO[List[Account]] = IO {
    DB.readOnly { implicit session =>
      sql"""
        SELECT
          account_id,
          account_code,
          account_name,
          account_type,
          is_summary_account,
          display_order,
          is_aggregation_target,
          balance,
          created_at,
          updated_at
        FROM accounts
        ORDER BY display_order, account_code
      """
        .map(rs => mapToAccount(rs))
        .list
        .apply()
    }
  }

  override def findByCode(accountCode: AccountCode): IO[Option[Account]] = IO {
    DB.readOnly { implicit session =>
      sql"""
        SELECT
          account_id,
          account_code,
          account_name,
          account_type,
          is_summary_account,
          display_order,
          is_aggregation_target,
          balance,
          created_at,
          updated_at
        FROM accounts
        WHERE account_code = ${accountCode.value}
      """
        .map(rs => mapToAccount(rs))
        .single
        .apply()
    }
  }

  override def save(account: Account): IO[Account] = IO {
    DB.localTx { implicit session =>
      val now = LocalDateTime.now()
      val generatedId = sql"""
        INSERT INTO accounts (
          account_code,
          account_name,
          account_type,
          is_summary_account,
          display_order,
          is_aggregation_target,
          balance,
          created_at,
          updated_at
        ) VALUES (
          ${account.accountCode.value},
          ${account.accountName},
          ${account.accountType.toString.toLowerCase},
          ${account.isSummaryAccount},
          ${account.displayOrder},
          ${account.isAggregationTarget},
          ${account.balance.value},
          $now,
          $now
        )
      """
        .updateAndReturnGeneratedKey
        .apply()

      account.copy(
        accountId = Some(generatedId),
        createdAt = now,
        updatedAt = now
      )
    }
  }

  override def update(account: Account): IO[Account] = IO {
    DB.localTx { implicit session =>
      val now = LocalDateTime.now()
      sql"""
        UPDATE accounts SET
          account_name = ${account.accountName},
          account_type = ${account.accountType.toString.toLowerCase},
          is_summary_account = ${account.isSummaryAccount},
          display_order = ${account.displayOrder},
          is_aggregation_target = ${account.isAggregationTarget},
          balance = ${account.balance.value},
          updated_at = $now
        WHERE account_code = ${account.accountCode.value}
      """
        .update
        .apply()

      account.copy(updatedAt = now)
    }
  }

  override def delete(accountCode: AccountCode): IO[Boolean] = IO {
    DB.localTx { implicit session =>
      val deleted = sql"""
        DELETE FROM accounts WHERE account_code = ${accountCode.value}
      """
        .update
        .apply()
      deleted > 0
    }
  }

  private def mapToAccount(rs: WrappedResultSet): Account =
    Account(
      accountId = Some(rs.long("account_id")),
      accountCode = AccountCode.unsafeFrom(rs.string("account_code")),
      accountName = rs.string("account_name"),
      accountType = AccountType.fromString(rs.string("account_type")).getOrElse(AccountType.Asset),
      isSummaryAccount = rs.boolean("is_summary_account"),
      displayOrder = rs.int("display_order"),
      isAggregationTarget = rs.boolean("is_aggregation_target"),
      balance = Money(rs.bigDecimal("balance")),
      createdAt = rs.localDateTime("created_at"),
      updatedAt = rs.localDateTime("updated_at")
    )

File: app/rust/accounting-service/financial_accounting/src/infrastructure/persistence/repositories/account_repository_impl.rs
use async_trait::async_trait;
use sqlx::PgPool;

use crate::application::ports::output::account_repository::AccountRepository;
use crate::domain::account::Account;

/// 勘定科目リポジトリの実装（Output Adapter）
pub struct AccountRepositoryImpl {
    pool: PgPool,
}

impl AccountRepositoryImpl {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl AccountRepository for AccountRepositoryImpl {
    async fn find_all(&self) -> Result<Vec<Account>, Box<dyn std::error::Error>> {
        let accounts = sqlx::query_as::<_, Account>(
            r#"
            SELECT
                "勘定科目ID", "勘定科目コード", "勘定科目名", "勘定科目カナ",
                "勘定科目種別"::TEXT as "勘定科目種別",
                "合計科目", "BSPL区分", "取引要素区分", "費用区分",
                "表示順序", "集計対象", "課税取引コード", "残高",
                "作成日時", "更新日時"
            FROM "勘定科目マスタ"
            ORDER BY "勘定科目コード"
            "#,
        )
        .fetch_all(&self.pool)
        .await?;

        Ok(accounts)
    }

    async fn find_by_code(
        &self,
        code: &str,
    ) -> Result<Option<Account>, Box<dyn std::error::Error>> {
        let account = sqlx::query_as::<_, Account>(
            r#"
            SELECT
                "勘定科目ID", "勘定科目コード", "勘定科目名", "勘定科目カナ",
                "勘定科目種別"::TEXT as "勘定科目種別",
                "合計科目", "BSPL区分", "取引要素区分", "費用区分",
                "表示順序", "集計対象", "課税取引コード", "残高",
                "作成日時", "更新日時"
            FROM "勘定科目マスタ"
            WHERE "勘定科目コード" = $1
            "#,
        )
        .bind(code)
        .fetch_optional(&self.pool)
        .await?;

        Ok(account)
    }

    async fn create(&self, account: Account) -> Result<Account, Box<dyn std::error::Error>> {
        let created = sqlx::query_as::<_, Account>(
            r#"
            INSERT INTO "勘定科目マスタ" (
                "勘定科目コード", "勘定科目名", "勘定科目カナ", "勘定科目種別",
                "合計科目", "BSPL区分", "取引要素区分", "費用区分",
                "表示順序", "集計対象", "課税取引コード", "残高"
            ) VALUES ($1, $2, $3, $4::account_type, $5, $6, $7, $8, $9, $10, $11, $12)
            RETURNING
                "勘定科目ID", "勘定科目コード", "勘定科目名", "勘定科目カナ",
                "勘定科目種別"::TEXT as "勘定科目種別",
                "合計科目", "BSPL区分", "取引要素区分", "費用区分",
                "表示順序", "集計対象", "課税取引コード", "残高",
                "作成日時", "更新日時"
            "#,
        )
        .bind(&account.account_code)
        .bind(&account.account_name)
        .bind(&account.account_name_kana)
        .bind(&account.account_type)
        .bind(account.is_summary_account)
        .bind(&account.bspl_type)
        .bind(&account.transaction_element_type)
        .bind(&account.expense_type)
        .bind(account.display_order)
        .bind(account.is_aggregation_target)
        .bind(&account.tax_code)
        .bind(account.balance)
        .fetch_one(&self.pool)
        .await?;

        Ok(created)
    }

    async fn update(&self, account: Account) -> Result<Account, Box<dyn std::error::Error>> {
        let updated = sqlx::query_as::<_, Account>(
            r#"
            UPDATE "勘定科目マスタ"
            SET
                "勘定科目名" = $2,
                "勘定科目カナ" = $3,
                "勘定科目種別" = $4::account_type,
                "合計科目" = $5,
                "BSPL区分" = $6,
                "取引要素区分" = $7,
                "費用区分" = $8,
                "表示順序" = $9,
                "集計対象" = $10,
                "課税取引コード" = $11,
                "残高" = $12,
                "更新日時" = CURRENT_TIMESTAMP
            WHERE "勘定科目コード" = $1
            RETURNING
                "勘定科目ID", "勘定科目コード", "勘定科目名", "勘定科目カナ",
                "勘定科目種別"::TEXT as "勘定科目種別",
                "合計科目", "BSPL区分", "取引要素区分", "費用区分",
                "表示順序", "集計対象", "課税取引コード", "残高",
                "作成日時", "更新日時"
            "#,
        )
        .bind(&account.account_code)
        .bind(&account.account_name)
        .bind(&account.account_name_kana)
        .bind(&account.account_type)
        .bind(account.is_summary_account)
        .bind(&account.bspl_type)
        .bind(&account.transaction_element_type)
        .bind(&account.expense_type)
        .bind(account.display_order)
        .bind(account.is_aggregation_target)
        .bind(&account.tax_code)
        .bind(account.balance)
        .fetch_one(&self.pool)
        .await?;

        Ok(updated)
    }
}
